docs/proposals/20210323-etcdadm-bootstrap-provider.md

```diff
@@ -0,0 +1,556 @@
+- Once the EtcdCluster is Initialized, the etcd cluster controller will scale up the cluster to add as many members as required to match EtcdCluster.Spec.Replicas field
+- At the end of every reconciliation loop, the etcd cluster controller will check if desired number of members are created by
+    * Getting Machines owned by EtcdCluster
+    * Running healthcheck for each Machine with the etcd `/health` endpoint at `https://client-url:2379/health`
+- Once all members are ready, EtcdCluster will set the endpoints on EtcdCluster.Status.Endpoints field, and set EtcdCluster.Status.Ready to true
+- The CAPI cluster controller will have a new `reconcileEtcdCluster` function, that will check if EtcdCluster is ready, and when it is ready it will unpause the controlplane by deleting the "paused" annotation.
+- The controlplane controller starts provisioning once it is unpaused. It will first check if the Cluster spec contains the managedExternalEtcd ref. If the reference exists, it will get the endpoint from the referred etcdCluster object's status field. It will then start the provisoning of the controlplane cluster using this etcd cluster.
+
+###### Upgrade or member replacement use case
+
+- The EtcdCluster.Spec.Replicas field will be immutable. This is to protect the etcd cluster from losing quorum by unnecessarily adding/removing healthy members.
+- An etcd member will need to be replaced by the etcd controller in two cases:
+  - Member is found unhealthy
+  - During an upgrade
+- While replacing an unhealthy etcd member, it is important to first remove the member and only then replace it by adding a new one. This ensures the cluster stays stable. [This Etcd doc](https://etcd.io/docs/v3.4/faq/#should-i-add-a-member-before-removing-an-unhealthy-member) explains why removing an unhealthy member is important in depth. Etcd's latest versions allow adding new members as learners so it doesn't affect the quorum size, but that feature is in beta. Also, adding a learner is a two-step process, where you first add a member as learner and then promote it, and etcdadm currently doesn't support it. So we can include this in a later version once the learner feature becomes GA and etcdadm incorporates the required changes.
+- We can reuse this same "scale down" function when removing a healthy member during an etcd upgrade. So even upgrade will follow the same pattern of removing the member first and then adding a new one with the latest spec.
+- Once the upgrade/machine replacement is completed, the etcd controller will again update the EtcdCluster.Status.Endpoint field. The controlplane controller will rollout new Machines that use the updated etcd endpoints for the apiserver.
+
+###### Periodic etcd member healthchecks
+- The controller should do healthchecks of the etcd members in a loop at a certain predeterminted interval (10s).
```

Comment:
Let's clarify exactly where this controller runs. Is that external to cluster? Also, we may want to add peer-to-peer health checks (via each agent running on the node), but this can be added later.

Reply:
yes the controller is external to the etcd cluster...currently no agent runs on the Machine resource in CAPI and so we can't perform peer-to-peer healthchecks, so I can add it in the future work section